lib/explorer/polars_backend/data_frame.ex

```diff
@@ -821,4 +822,43 @@ defmodule Explorer.PolarsBackend.DataFrame do
   def inspect(df, opts) do
     Explorer.Backend.DataFrame.inspect(df, "Polars", n_rows(df), opts)
   end
+
+  # Helpers
+
+  defp maybe_swap_args(%Explorer.Backend.LazySeries{op: op, args: args}, df) do
+    maybe_swapped_args =
+      case op do
+        :add ->
+          case List.pop_at(args, -1) do
+            {%Explorer.PolarsBackend.Series{} = last, rest} ->
+              case Explorer.PolarsBackend.Native.s_dtype(last) do
+                {:ok, "date"} ->
+                  [last | rest]
+
+                _ ->
+                  args
+              end
+
+            {%Explorer.Backend.LazySeries{op: :column, args: [col_name]} = last, rest} ->
```

Comment:
Perhaps the best place to swap args is in expression.ex. In there we convert all lazy frame operations into Polars expressions, so you can override `add`, `multiply`, and friends in there to both swap and post-cast.

Reply:
I tried there initially, but lazy series didn't seem to contain the dtype information I needed to make decisions about whether to swap, cast, etc.

E.g. if we add this custom `to_expr` clause for add:

```elixir
  def to_expr(%LazySeries{op: :add, args: [left, right]}) do
    left |> IO.inspect(label: :left, width: 120)
    right |> IO.inspect(label: :right, width: 120)

    Native.expr_add(to_expr(left), to_expr(right))
  end
```

We see the following after running `df1 = DF.mutate(df, add1: sub + aug_20)`:

```elixir
left: %Explorer.Backend.LazySeries{op: :column, args: ["aug_20"], aggregation: false}
right: %Explorer.Backend.LazySeries{op: :column, args: ["sub"], aggregation: false}
```

I don't know the dtypes of those args, so I don't know if I should swap or not.

Reply:
GAH. Does it work in expression.rs? Can we fetch the dtype there and swap there?

If not, we would need to start storing the dtype then... it is more work though, so it probably makes sense to break this apart, add the dtype field to lazy series, and then come back to this. If we indeed can't do it in Rust, then we should merge a subset of this PR that doesn't require the arg swapping, and work on the other fix later.

I guess there is always the option of fixing Polars too.

Reply:
> Does it work in expression.rs? Can we fetch the dtype there and swap there?

I don't _think_ so? I'll try to check more thoroughly later today tho. Polars expressions weren't super well documented so I'll have to read some source code.

> If we indeed can't do it in Rust, then we should merge a subset of this PR that doesn't require the arg swapping, and work on the other fix later.

Do you mean supporting `Explorer.Series.add(duration, date)` a letting a lazy expression possibly panic? Or dropping `duration + date` support altogether until we can address the issue?

> I guess there is always the option of fixing Polars too.

You read my mind...

Reply:
> Do you mean supporting Explorer.Series.add(duration, date) a letting a lazy expression possibly panic? 

I meant adding a field called dtype to `Explorer.Backend.LazySeries`. If we do this though, I would make it so we change all Explorer.Backends.Series callbacks to receive the output dtype as argument (or an `out_series`, which has the correct result type, like we do for dataframes).

Reply:
Sorry I meant what should we do for this PR? Definitely coming back later and adding the dtype info we need to make decisions sounds good.

Reply:
Actually, let's keep the arg swap in `Explorer.Series` for now. It is not ideal but it is semantically fine. We can add a TODO to them.

For the integer vs duration, which requires the post swap, we can pre-cast the integer to duration, right?

Reply:
> Actually, let's keep the arg swap in Explorer.Series for now. It is not ideal but it is semantically fine. We can add a TODO to them.

👍 

> For the integer vs duration, which requires the post cast, we can pre-cast the integer to duration, right?

No, `duration * duration` isn't supported. I think only way to do it is post-cast.

Reply:
I honestly think opening up a PR in Polars might be the least painful way accomplish this 😅